### PR TITLE
seamless authentication from CLI (no more token pasting!) addresses DSDEEPB-355

### PR DESCRIPTION
**Pull request by kcibul**
* added Google Application Default Credentials
* added --insecure flag to connect to servers without valid SSL certificates
* refactored into packges
* added setuptools install
* created installation script for end-users
* cleaned up installation to avoid users putting junk in their path
* better error handling
* added more informative message if gcloud credentials could not be used

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,14 @@
 # firecloud-cli
 Command line tools for Firecloud
+
+## Installation
+The FireCloud CLI requires:
+* Python 2.7+
+* Virtualenv (run `sudo pip install virtualenv`)
```
Comment (bradtaylor) [inline README.md]: This assumes a linux/OSX environment. Should we provide a separate installation script for users like Hussein who use Windows?

File: firecloudcli/agora.py [hunk 2]
```diff
@@ -89,45 +94,65 @@ def get_user_synopsis():
...
+def httpRequest(baseUrl, path, insecureSsl, method, requestBody, expectedReturnStatus):
+    http = httplib2.Http(".cache", disable_ssl_certificate_validation=(not insecureSsl))
+
+    # get the credentials for the google proxy using the application default, which
+    # is described at 
+    # https://developers.google.com/identity/protocols/application-default-credentials#howtheywork
+    #
+    # Most often, this is obtained from the users current 'gcloud' credentials
+    credentials = GoogleCredentials.get_application_default()
+    http = credentials.authorize(http)
```
Comment (bradtaylor) [inline firecloudcli/agora.py]: Can this throw errors? Does it also need to be in the try?

File: firecloudcli/agora.py [hunk 3]
```diff
@@ -89,45 +94,65 @@ def get_user_synopsis():
...
+        )
+    except AccessTokenRefreshError as atre:
+        print "******************************************************************************"
+        print "* ERROR: Unable to access your google credentials.                           *"
+        print "*                                                                            *"
+        print "* This is most often caused by either                                        *"
+        print "*     (a) not having gcloud installed (see https://cloud.google.com/sdk/)    *"
+        print "*     (b) gcloud is not logged in (run 'gcloud auth login')                  *"
+        print "******************************************************************************"
+        exit(1)
```
Comment (bradtaylor) [inline firecloudcli/agora.py]: There is a fail method that takes a string message, prints it, and exits. Do we want to use that (either build the message as on line 104 below or triple quote this multi-line message?)

File: firecloudcli/agora.py [hunk 4]
```diff
@@ -89,45 +94,65 @@ def get_user_synopsis():
...
+        raise e
+
     if response.status != expectedReturnStatus:
         message = ("[ERROR] Agora HTTP request failed\n"
                    "Request URL: " + path + "\n"
                    "Request body:\n"
                    + str(requestBody) + "\n"
                    "Response:\n"
-                   + str(response.status) + " " + response.reason + " " + data
+                   + str(response.status) + " " + response.reason + " " + content
```
Comment (bradtaylor) [inline firecloudcli/agora.py]: What is content here and where is it coming from? I do not see anything about "content" in the docs to httplib

File: firecloudcli/agora.py [hunk 5]
```diff
@@ -8,25 +8,30 @@
...
-# Must find a way to programatically get an Oath token from the openAM server.
-#   Currently, assumes you just grabbed the token and are passing it in as an
-#   argument. Really hackish.
 # 
 ################################################################################
 
 
+from oauth2client.client import GoogleCredentials
+from oauth2client.client import AccessTokenRefreshError
+import httplib2
```
Comment (bradtaylor) [inline firecloudcli/agora.py]: what does httplib2 provide that httplib does not?

File: README.md [hunk 6]
```diff
@@ -1,2 +1,14 @@
...
+* Google Cloud SDK (see https://cloud.google.com/sdk/)
+
+To install run `./install.sh` and follow the instructions to add ~./firecloud-cli/venv/bin to your default path
+
+## Usage
+
+After installation run `agora` for usage
```
Comment (bradtaylor) [inline README.md]: Should we tell them to run "firecloud" instead?

File: firecloudcli/main.py [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import agora
+
+def main():
+    # for now, just call the agora endpoint
+    agora.main()
```
Comment (bradtaylor) [inline firecloudcli/main.py]: Larger discussion- should we call agora something else? It seems like the name of the methods repo github might not be the most user-friendly thing here.

File: firecloudcli/agora.py [hunk 8]
```diff
@@ -161,21 +187,25 @@ def list_entities(args):
...
+
+def main():
     # The main argument parser
     parser = ArgumentParser(description="CLI for accessing the AGORA methods repository. Currently only handles method push")
     
     # Core application arguments
-    parser.add_argument('-a', '--auth', dest='auth', action='store', help='Oath token key=value pair for passing in request cookies')
-    parser.add_argument('-u', '--url', dest='agoraUrl', default='agora.dsde-staging.broadinstitute.org', action='store', help='Agora location. Default is agora.dsde-staging.broadinstitute.org')
+    parser.add_argument('-u', '--url', dest='agoraUrl', default='https://agora.dsde-dev.broadinstitute.org', action='store', help='Agora location. Default is https://agora.dsde-dev.broadinstitute.org')
+    parser.add_argument('-k', '--insecure', dest='insecureSsl', default='False', action='store_true', help='use insecure ssl (allow self-signed certificates)')
```
Comment (bradtaylor) [inline firecloudcli/agora.py]: request for education- what does this mean and when would I use it?

File: firecloudcli/agora.py [hunk 9]
```diff
@@ -89,45 +94,65 @@ def get_user_synopsis():
...
+        print "*     (a) not having gcloud installed (see https://cloud.google.com/sdk/)    *"
+        print "*     (b) gcloud is not logged in (run 'gcloud auth login')                  *"
+        print "******************************************************************************"
+        exit(1)
     except Exception as e:
-        raise httplib.HTTPException("Could not connect to {0}{1} with method {2}".format(baseUrl,path, method))
-    response = conn.getresponse()
-    data = response.read()
+        print "Could not connect to {0}{1} with method {2}".format(baseUrl,path, method)
+        raise e
```
Comment (bradtaylor) [inline firecloudcli/agora.py]: should we raise this or fail() as elsewhere? I don't have much experience with python error-handling best practices. Should we be raising custom exceptions elsewhere? I just want to make sure errors are handled consistently throughout the script

File: setup.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+      'License :: OSI Approved :: BSD License',
+      'Programming Language :: Python',
+      'Development Status :: 4 - Beta',
+      'Intended Audience :: Developers',
+      'Natural Language :: English'
+  ],
+  entry_points = {
+        'console_scripts': [
+            'firecloud = firecloudcli.main:main',
+            'agora = firecloudcli.agora:main'
```
Comment (bradtaylor) [inline setup.py]: Should these both be here considering firecloudcli.main -> agora.main?

File: install.sh [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    echo
+    echo "------------------------------"
+    echo "ERROR... aborting installation"
+    echo "------------------------------"
+    exit $rv
+}
+
+trap "cleanup" INT TERM EXIT
+
+rm -rf ~/.firecloud-cli && \
```
Comment (bradtaylor) [inline install.sh]: Why are these commands anded together vs being packed in a function like "cleanup" above?

**Comment by bradtaylor** [comment]
This is very neat. Not a lot of code to get your requests authorized. Also, I've never worked with setuptools, python virtual environments, or traps in bash. So I learned something :)

File: firecloudcli/agora.py (hunk 5 above)
Comment (kcibul) [inline firecloudcli/agora.py]: yep -- that's exactly it.  It probably does new things also, but the main motivator was that GoogleCredentials.authorize works with httplib2

File: firecloudcli/agora.py (hunk 2 above)
Comment (kcibul) [inline firecloudcli/agora.py]: I don't think it does, but good idea to move it inside anyways, I'll do that

File: README.md (hunk 1 above)
Comment (kcibul) [inline README.md]: it's a good question, I'll update the readme to say these instructions are just for Linux/OSX and leave it for the future to support windows users.  I don't think we should hold this up for that.  Let's let hussein figure it out ;)

File: firecloudcli/agora.py (hunk 3 above)
Comment (kcibul) [inline firecloudcli/agora.py]: yep -- sounds good

File: README.md (hunk 6 above)
Comment (kcibul) [inline README.md]: i was saving that for when we can refactor the cli design (e.g. is it firecloud methods blah blah).  when we do that, I 100% agree

File: firecloudcli/main.py (hunk 7 above)
Comment (kcibul) [inline firecloudcli/main.py]: yep -- larger discussion ;)  But I agree.  I think the right time to do this is when we rehome to the orchestration API & design the actual cli (as above... is it "firecloud methods <etc>" for this stuff.

File: firecloudcli/agora.py (hunk 8 above)
Comment (kcibul) [inline firecloudcli/agora.py]: I hope the parameter describes it well "use insecure ssl (allow self-signed certificates)'"

Self signed certs are those ones that when you visit the http server in the browser it says like "This site is not safe" and you have to check the box to continue.  This flag is the equivalent of checking that box

**Comment by kcibul** [comment]
comments addressed, rebased & squashed

**Pull request merged**